libGeoIP/GeoIP.c

```diff
@@ -1359,39 +1359,38 @@ unsigned int
 _GeoIP_seek_record_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl) {
     int depth;
     char paddr[ADDR_STR_LEN];
-    unsigned int x;
+    size_t x;
     unsigned char stack_buffer[2 * MAX_RECORD_LENGTH];
     const unsigned char *buf = (gi->cache == NULL) ? stack_buffer : NULL;
-    unsigned int offset = 0;
+    size_t offset = 0;
 
     const unsigned char *p;
-    int j;
-    int fno = fileno(gi->GeoIPDatabase);
+    size_t j;
 
-    unsigned int record_pair_length = gi->record_length * 2;
+    const size_t record_pair_length = gi->record_length * 2;
 
     _check_mtime(gi);
     if (GeoIP_teredo(gi)) {
         __GEOIP_PREPARE_TEREDO(&ipnum);
     }
     for (depth = 127; depth >= 0; depth--) {
-        unsigned int byte_offset = record_pair_length * offset;
+        size_t byte_offset = record_pair_length * offset;
         if (byte_offset > gi->size - record_pair_length) {
             /* The pointer is invalid */
             break;
         }
         if (gi->cache == NULL && gi->index_cache == NULL) {
             /* read from disk */
-            if (pread(
-                    fno, stack_buffer, record_pair_length, (long)byte_offset) !=
-                record_pair_length) {
+            int fno = fileno(gi->GeoIPDatabase);
```

Comment:
Although this improves the performance when caching is being used, it now means that this will get called dozens of extra times per lookup otherwise.

Reply:
Have you measured how much performance is gained by caching the `fileno()` result? This cached mode wastes a **lot** of performance by doing so many `pread()` system calls, and it wastes a **lot** of memory by duplicating data which is already in the kernel's page cache. I guess that this overhead is many orders of magnitude larger than those extra `fileno()` calls.
I don't know why anybody would ever use this caching, when memory mapped access is the fastest way to do it and has no disadvantages.
Another general problem here is that you're using the "stdio" buffered I/O API and then circumvent the buffering by using `pread()`. Of course, when seeking so much, stdio's naive buffering is of no use, but why use this API in the first place?